fvm/evm/handler/handler.go

```diff
@@ -0,0 +1,414 @@
+}
+
+// LastExecutedBlock returns the last executed block
+func (h ContractHandler) LastExecutedBlock() *types.Block {
+	block, err := h.blockchain.LatestBlock()
+	handleError(err)
+	return block
+}
+
+// Run runs an rlpencoded evm transaction and
+// collects the gas fees and pay it to the coinbase address provided.
+func (h ContractHandler) Run(rlpEncodedTx []byte, coinbase types.Address) {
+
+	// step 1 - transaction decoding
+	err := h.backend.MeterComputation(environment.ComputationKindRLPDecoding, uint(len(rlpEncodedTx)))
+	handleError(err)
+
+	tx := gethTypes.Transaction{}
+	err = tx.DecodeRLP(
+		rlp.NewStream(
+			bytes.NewReader(rlpEncodedTx),
+			uint64(len(rlpEncodedTx))))
+	handleError(err)
+
+	// step 2 - run transaction
+	h.checkGasLimit(types.GasLimit(tx.Gas()))
+
+	ctx := h.getBlockContext()
+	ctx.GasFeeCollector = coinbase
+	blk, err := h.emulator.NewBlockView(ctx)
+	handleError(err)
+
+	res, err := blk.RunTransaction(&tx)
+	h.meterGasUsage(res)
+	handleError(err)
+
+	// step 3 - process block and emit events
+	txHash := tx.Hash()
+	h.appendTxHashToBlockDraft(txHash)
+
+	h.emitEvent(types.NewTransactionExecutedEvent(
+		h.getBlockDraftHeight(),
+		rlpEncodedTx,
+		txHash,
+		res,
+	))
+	h.updateBlockDraftStateRoot(res.StateRootHash)
+	h.commitBlockDraft()
+}
+
+// this method has been separated out to make NewContractHandler lighter
+// and lazy load these operations if needed
+func (h *ContractHandler) setupNewBlockDraft() {
+	lastExecutedBlock, err := h.blockchain.LatestBlock()
+	handleError(err)
+
+	parentHash, err := lastExecutedBlock.Hash()
+	if err != nil {
+		// this is a fatal error
+		panic(err)
+	}
+	h.newBlockDraft = &types.Block{
+		Height:            lastExecutedBlock.Height + 1,
+		ParentBlockHash:   parentHash,
+		TotalSupply:       lastExecutedBlock.TotalSupply,
+		TransactionHashes: make([]gethCommon.Hash, 0),
+	}
+}
+
+func (h *ContractHandler) updateBlockDraftStateRoot(stateRoot gethCommon.Hash) {
+	if h.newBlockDraft == nil {
+		h.setupNewBlockDraft()
+	}
+	h.newBlockDraft.StateRoot = stateRoot
+}
+
+func (h *ContractHandler) updateBlockDraftTotalSupply(newValue uint64) {
+	if h.newBlockDraft == nil {
+		h.setupNewBlockDraft()
+	}
+	h.newBlockDraft.TotalSupply = newValue
+}
+
+func (h *ContractHandler) getBlockDraftTotalSupply() uint64 {
+	if h.newBlockDraft == nil {
+		h.setupNewBlockDraft()
+	}
+	return h.newBlockDraft.TotalSupply
+}
+
+func (h *ContractHandler) appendTxHashToBlockDraft(txHash gethCommon.Hash) {
+	if h.newBlockDraft == nil {
+		h.setupNewBlockDraft()
+	}
+	h.newBlockDraft.TransactionHashes = append(h.newBlockDraft.TransactionHashes, txHash)
+}
+
+func (h *ContractHandler) getBlockDraftHeight() uint64 {
+	if h.newBlockDraft == nil {
+		h.setupNewBlockDraft()
+	}
+	return h.newBlockDraft.Height
+}
+
+func (h *ContractHandler) commitBlockDraft() {
+	err := h.blockchain.AppendBlock(h.newBlockDraft)
+	handleError(err)
+	h.emitEvent(types.NewBlockExecutedEvent(h.newBlockDraft))
```

Comment:
I wonder if this could cause exceeding computation limit. I might be nitpicking, but could there be a case where the tx gas limit is the same as computation limit, so the checkGasLimit passes but then emitting the event adds computation usage and causes the transaction to fail. I guess it could be ok since it would revert the transaction, but I rather ask than not.

Reply:
emitting events are metered through the FVM backend, so they would fail similarly to how cadence transactions fail.